### PR TITLE
[rllib] Flexible multi-agent replay modes and replay_sequence_length

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

This adds a "replay_mode" option for multiagent training. When replay_mode=independent (current state of things), RLlib provides no guarantees on whether all the agent experiences for a particular timestep are present in the same training batch together. When replay_mode=lockstep, RLlib will replay all the agent experiences at a particular timestep together in the batch. Some things like prioritized replay and SGD minibatches are difficult to tricky to implement in lockstep mode, so those are currently unsupported.

Implementation:
- The replay buffer has been refactored to store MultiAgentSampleBatch as items. This enables storing sequences as well as lockstep mode.
- Add timeslice() utils to SampleBatch and MultiAgentSampleBatch to support sequence / lockstep replay modes.

This PR also adds a new "replay_sequence_length" option that can work in either mode. It's intended for future use with training RNN and attention models.

## Related issue number

Part of https://github.com/ray-project/ray/issues/7341

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [WIP] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: rllib/execution/replay_buffer.py [hunk 1]
```diff
@@ -1,31 +1,34 @@
...
-    MultiAgentBatch
+from ray.rllib.policy.sample_batch import SampleBatch, MultiAgentBatch, \
+    DEFAULT_POLICY_ID
 from ray.rllib.utils.annotations import DeveloperAPI
-from ray.rllib.utils.compression import unpack_if_needed
 from ray.util.iter import ParallelIteratorWorker
 from ray.rllib.utils.timer import TimerStat
 from ray.rllib.utils.window_stat import WindowStat
 
+# Constant that represents all policies in lockstep replay mode.
```
Comment (ericl) [inline rllib/execution/replay_buffer.py:18]: The most significant changes are to this file.

File: rllib/agents/trainer.py [hunk 2]
```diff
@@ -347,8 +347,19 @@
...
         "observation_fn": None,
+        # When replay_mode=lockstep, RLlib will replay all the agent
+        # transitions at a particular timestep together in a batch. This allows
+        # the policy to implement differentiable shared computations between
+        # agents it controls at that timestep. When replay_mode=independent,
+        # transitions are replayed independently per policy.
+        "replay_mode": "independent",
     },
 
+    # === Replay Settings ===
```
Comment (sven1977) [inline rllib/agents/trainer.py:358]: This is very cool. We should be able to implement R2D2 then just setting this to >1.

**Review by sven1977 — APPROVED** [review]
LGTM. Thanks for cleaning up the buffers.

**Comment by ericl** [comment]
I believe test_rollout.py is flaking in master as well due to timeouts (@sven1977 )

**Pull request merged**